### PR TITLE
[dynamo 3.11] remap dup/rotate to copy/swap

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94243
* #94102
* #94101
* #94100
* #94099
* #94098
* __->__ #93988
* #93987
* #93986

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/bytecode_transformation.py [hunk 1]
```diff
@@ -63,6 +63,33 @@ def create_jump_absolute(target):
...
+        return []
+
+    if sys.version_info >= (3, 11):
+        # rotate can be expressed as a sequence of swap operations
+        # e.g. rotate 3 is equivalent to swap 3, swap 2
+        return [create_instruction("SWAP", i) for i in range(n, 1, -1)]
+
+    # ensure desired rotate function exists
+    if sys.version_info < (3, 8) and n >= 4:
+        raise AttributeError(f"rotate {n} not supported for Python < 3.8")
```
Comment (albanD) [inline torch/_dynamo/bytecode_transformation.py:84]: This used to work on all python versions. Why are we restricting it here?

Comment (albanD) [reply torch/_dynamo/bytecode_transformation.py:84]: Ho this is caught by the caller... That feel bad, but ok.

**Review by albanD — APPROVED** [review]

File: torch/_dynamo/bytecode_transformation.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/bytecode_transformation.py:84]: I'll make a comment to clarify

File: torch/_dynamo/codegen.py [hunk 2]
```diff
@@ -259,24 +263,21 @@ def load_function_name(self, fn_name, num_on_stack=0):
...
-                create_instruction("BUILD_TUPLE", n),
-                self._create_load_const(rot_n_helper(n)),
-                create_instruction("ROT_TWO"),
-                create_instruction("CALL_FUNCTION_EX", 0),
-                create_instruction("UNPACK_SEQUENCE", n),
-            ]
+        try:
+            return create_rot_n(n)
+        except AttributeError:
+            # desired rotate bytecode doesn't exist, generate equivalent bytecode
```
Comment (mlazos) [inline torch/_dynamo/codegen.py:269]: Could we just always generate this equivalent bytecode?

**Review by mlazos — APPROVED** [review]

File: torch/_dynamo/codegen.py (hunk 2 above)
Comment (williamwen42) [reply torch/_dynamo/codegen.py:269]: It's not as efficient.

**Pull request closed without merging**